core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,284 @@
+package pureconfig
+
+import java.io.File
+import java.net.URL
+import java.nio.file.Path
+
+import scala.reflect.ClassTag
+
+import com.typesafe.config._
+import pureconfig.ConfigReader.Result
+import pureconfig.backend.{ ConfigFactoryWrapper, PathUtil }
+import pureconfig.error.{ ConfigReaderException, ConfigReaderFailures }
+
+/**
+ * A representation of a source from which `ConfigValue`s can be loaded, such as a file or a URL.
+ *
+ * A source allows users to load configs from this source as any type for which a `ConfigReader` is
+ * available. Raw configs can also be retrieved as a `ConfigValue`, a `ConfigCursor` or a
+ * `FluentConfigCursor`. Before using any of the loading methods described, Users can opt to focus
+ * on a specific part of a config by specifying a namespace.
+ *
+ * All config loading methods are lazy and defer resolution of references until needed.
+ */
+trait ConfigSource {
+
+  /**
+   * Retrieves a `ConfigValue` from this source.
+   *
+   * @return a `ConfigValue` retrieved from this source.
+   */
+  def value: Result[ConfigValue]
```

Comment:
I think that both the name and the signature here should reflect that it performs side effects:
```scala
def getValue(): Result[Config]
```

Reply:
I added empty parenthesis to the signature and opted to keep the function name the same (mostly a personal preference; I'm not very fond of `get` prefixes in methods as it doesn't convey a lot of additional information).